code/modules/lootpanel/loot_procs.dm

```diff
@@ -0,0 +1,171 @@
+/// Converts the contents of the tile into search objects
+/datum/lootpanel/proc/convert_tile_contents(list/slice)
+	var/list/found = list()
+
+	for(var/atom/thing as anything in slice)
+		if(thing.mouse_opacity == MOUSE_OPACITY_TRANSPARENT)
+			continue
+		if(thing.IsObscured())
+			continue
+		if(thing.invisibility > user.see_invisible)
+			continue
+
+		var/ref = REF(thing)
+		if(contents[ref])
+			continue
+
+		RegisterSignals(thing, list(
+			COMSIG_MOVABLE_MOVED,
+			COMSIG_QDELETING,
+			), PROC_REF(on_item_moved))
+
+		var/datum/search_object/item = new(user, thing)
+		found[ref] = item
+
+	return found
+
+
+/// Helper function for removing a search object
+/datum/lootpanel/proc/delete_search_object(datum/search_object/item)
+	contents -= item.string_ref
+	qdel(item)
+
+
+/// UI helper for converting the associative list to a list of lists
+/datum/lootpanel/proc/get_contents()
+	var/list/items = list()
+
+	for(var/ref in contents)
+		var/datum/search_object/item = contents[ref]
+		if(!item.item_ref?.resolve())
+			contents -= ref
+			continue
+
+		UNTYPED_LIST_ADD(items, list(
+			"icon_state" = item.icon_state,
+			"icon" = item.icon,
+			"name" = item.name, 
+			"path" = item.path,
+			"ref" = item.string_ref, 
+		))
+	
+	return items
+
+
+/// Grabs an object from the contents. Validates the object and the user
+/datum/lootpanel/proc/grab(mob/user, list/params)
+	var/ref = params["ref"]
+	if(isnull(ref))
+		return FALSE
+
+	var/datum/search_object/obj = contents[ref]
+	if(isnull(obj))
+		return FALSE
+
+	var/turf/source_tile = search_turf_ref?.resolve()
+	if(isnull(source_tile))
+		return FALSE
+
+	if(!user.TurfAdjacent(source_tile))
+		stop_search()
+		return FALSE
+
+	var/atom/thing = obj.item_ref?.resolve()	
+	if(QDELETED(thing))
+		return FALSE
+
+	if(thing != source_tile && (!thing.Adjacent(user) || !locate(thing) in source_tile.contents))
```

Comment:
reminder that !thing in list is interpreted as (!thing) in list

Reply:
oh actually may be wrong here lemme see

Reply:
this is actually fine but I would appreciate it if you moved the ! to an outer () to avoid people seeing this and assuming it holds true for the other evil in precedence